### PR TITLE
fix(Core/Movement: Rewritten fleeing movement generator. Source: TrinityCore.

### LINKED ISSUE #5980: Feared mobs walking through closed doors
#### WHAT CLIENT DO YOU PLAY ON?

- [ ] enGB
- [X] enUS

##### FACTION

- [ ] Alliance
- [x] Horde

##### CONTENT PHASE:

- [ ] Generic
- [ ] 1-19
- [ ] 20-29
- [X] 30-39
- [ ] 40-49
- [ ] 50-59

Original report: https://github.com/chromiecraft/chromiecraft/issues/677

##### CURRENT BEHAVIOUR:

Inside scarlet monastery cathedral,
when a mob gets feared by any source
and the cathedral's door is closed,
if the mob walks towards the door
it will keep walking through and pull everything on the other side.

##### EXPECTED BLIZZLIKE BEHAVIOUR:

Feared mobs should not walk through closed doors.

##### STEPS TO REPRODUCE THE PROBLEM:

1. Enter Scarlet Monastery Cathedral
2. Progress until the closed door
3. Have any mob near the door and have it feared (warlock's or warrior's spells, or low hp)
4. Repeat from 3 until a mob's random fear movement makes it walk through the door

##### AC HASH/COMMIT:

https://github.com/chromiecraft/azerothcore-wotlk/commit/e50754d804987ec93792b6a4d113913ab018189b

##### OPERATING SYSTEM:

Ubuntu 20.04

##### MODULES:

- [mod-cfbg](https://github.com/azerothcore/mod-cfbg)
- [mod-duel-reset](https://github.com/azerothcore/mod-duel-reset)
- [mod-desertion-warnings](https://github.com/azerothcore/mod-desertion-warnings)
- [mod-ah-bot](https://github.com/azerothcore/mod-ah-bot)
- [mod-eluna-lua-engine](https://github.com/azerothcore/mod-eluna-lua-engine)
- [mod-server-auto-shutdown](https://github.com/azerothcore/mod-server-auto-shutdown)
- [lua-LevelUpReward](https://github.com/55Honey/Acore_LevelUpReward)
- [lua-CarbonCopy](https://github.com/55Honey/Acore_CarbonCopy)
- [mod-pvp-titles](https://github.com/azerothcore/mod-pvp-titles)

##### OTHER CUSTOMIZATIONS:

None.

##### SERVER:

ChromieCraft

I have tried to reproduce it on a local AC server (AC HASH/COMMIT: a9e3d8d):

WoWScrnShot_052121_050924
WoWScrnShot_052121_050959
WoWScrnShot_052121_051136

But the most I have achieved is for one of them to go through the door, without bringing the mobs from inside.

<bountysource-plugin>

---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/98598887-feared-mobs-walking-through-closed-doors?utm_campaign=plugin&utm_content=tracker%2F40032087&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F40032087&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### LINKED ISSUE #9314: WSG - Enemy Players Can Be Feared Into World Object
### What client do you play on?

enUS

### Faction

- [X] Alliance
- [X] Horde

### Content Phase:

- [X] Generic
- [ ] 1-19
- [ ] 20-29
- [ ] 30-39
- [ ] 40-49
- [ ] 50-59

### Current Behaviour
Original report: https://github.com/chromiecraft/chromiecraft/issues/2381

WoWScrnShot_112321_155235

If I remember correctly, my enemy was trying to hide there. He/she stood in some kind of corner.

image

In Warsong Gulch battleground you can fear an enemy inside the line-of-sight object in the middle of the battleground.

Tester: confirmed. After the victim stood in the indicated spot, fear made them clip right through the object:

https://user-images.githubusercontent.com/81782124/143240008-aea84d7d-40c7-4eb6-9ba6-e2938f238c9d.mp4

Related issues are https://github.com/azerothcore/azerothcore-wotlk/issues/5980 (for NPCs) and closed issue https://github.com/azerothcore/azerothcore-wotlk/issues/4769.

### Expected Blizzlike Behaviour

It should not be possible to fear an enemy inside the line-of-sight object in the middle of the Warsong Gulch battleground.

### Source

This is not controversial.

### Steps to reproduce the problem

1. Join WSG battleground
2. Fear an enemy next to the line-of-sight object in the middle of the battleground

### AC rev. hash/commit

https://github.com/chromiecraft/azerothcore-wotlk/commit/00dc369cb627d728c520cd597b31e9edf2c75eed

### Operating system

Windows 10

### Modules

- [mod-ah-bot](https://github.com/azerothcore/mod-ah-bot)
- [mod-cfbg](https://github.com/azerothcore/mod-cfbg)
- [mod-chat-transmitter](https://github.com/azerothcore/mod-chat-transmitter)
- [mod-chromie-xp](https://github.com/azerothcore/mod-chromie-xp)
- [mod-desertion-warnings](https://github.com/azerothcore/mod-desertion-warnings)
- [mod-duel-reset](https://github.com/azerothcore/mod-duel-reset)
- [mod-eluna-lua-engine](https://github.com/azerothcore/mod-eluna-lua-engine)
- [mod-ip-tracker](https://github.com/azerothcore/mod-ip-tracker)
- [mod-low-level-arena](https://github.com/azerothcore/mod-low-level-arena)
- [mod-multi-client-check](https://github.com/azerothcore/mod-multi-client-check)
- [mod-pvp-titles](https://github.com/azerothcore/mod-pvp-titles)
- [mod-pvpstats-announcer](https://github.com/azerothcore/mod-pvpstats-announcer)
- [mod-queue-list-cache](https://github.com/azerothcore/mod-queue-list-cache)
- [mod-server-auto-shutdown](https://github.com/azerothcore/mod-server-auto-shutdown)
- [lua-carbon-copy](https://github.com/55Honey/Acore_CarbonCopy)
- [lua-custom-worldboss](https://github.com/55Honey/Acore_CustomWorldboss)
- [lua-level-up-reward](https://github.com/55Honey/Acore_LevelUpReward)
- [lua-recruit-a-friend](https://github.com/55Honey/Acore_RecruitAFriend)
- [lua-send-and-bind](https://github.com/55Honey/Acore_SendAndBind)
- [lua-temp-announcements](https://github.com/55Honey/Acore_TempAnnouncements)
- [lua-zonecheck](https://github.com/55Honey/acore_Zonecheck)
- [lua-zone-debuff](https://github.com/55Honey/Acore_ZoneDebuff)

### Customizations

None

### Server

ChromieCraft

### PR DESCRIPTION
**Pull request by UltraNix**
Fixes #5980
Fixes #9314

## Issues Addressed:

- Closes #5980
- Closes #9314
- Closes https://github.com/chromiecraft/chromiecraft/issues/677
- Closes https://github.com/chromiecraft/chromiecraft/issues/2381

## Tests Performed:

- Tested ingame.

## How to Test the Changes:

Enter Scarlet Monastery Cathedral
Progress until the closed door
Have any mob near the door and have it feared (warlock's or warrior's spells, or low hp)
Repeat from 3 until a mob's random fear movement makes it walk through the door

Join WSG battleground
Fear an enemy next to the line-of-sight object in the middle of the battleground (more info: [here](https://github.com/chromiecraft/chromiecraft/issues/677))

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

File: src/server/game/Movement/MovementGenerators/FleeingMovementGenerator.cpp [hunk 1]
```diff
@@ -28,351 +28,185 @@
...
 {
-    if (!owner)
-        return false;
+}
 
-    const Map* _map = owner->GetMap();
+template<>
+void FleeingMovementGenerator<Player>::DoFinalize(Player* owner)
+{
+    owner->RemoveFlag(UNIT_FIELD_FLAGS, UNIT_FLAG_FLEEING);
```
Comment (Kitzunu) [inline src/server/game/Movement/MovementGenerators/FleeingMovementGenerator.cpp:52]:
```suggestion
    owner->RemoveUnitFlag(UNIT_FLAG_FLEEING);
```

File: src/server/game/Movement/MovementGenerators/FleeingMovementGenerator.cpp [hunk 2]
```diff
@@ -28,351 +28,185 @@
...
+    owner->StopMoving();
+}
 
-    x = owner->GetPositionX();
-    y = owner->GetPositionY();
-    z = owner->GetPositionZ();
+template<>
+void FleeingMovementGenerator<Creature>::DoFinalize(Creature* owner)
+{
+    owner->RemoveFlag(UNIT_FIELD_FLAGS, UNIT_FLAG_FLEEING);
```
Comment (Kitzunu) [inline src/server/game/Movement/MovementGenerators/FleeingMovementGenerator.cpp:60]: same here

File: src/server/game/Movement/MovementGenerators/FleeingMovementGenerator.cpp [hunk 3]
```diff
@@ -28,351 +28,185 @@
...
-    }
-
-    owner->AddUnitState(UNIT_STATE_FLEEING_MOVE);
-
-    Movement::MoveSplineInit init(owner);
-    init.MoveTo(x, y, z, true);
-    init.SetWalk(false);
-    init.Launch();
+    _path = nullptr;
+    owner->SetFlag(UNIT_FIELD_FLAGS, UNIT_FLAG_FLEEING);
```
Comment (Kitzunu) [inline src/server/game/Movement/MovementGenerators/FleeingMovementGenerator.cpp:39]:
```suggestion
    owner->SetUnitFlag(UNIT_FLAG_FLEEING);
```

**Review by Kitzunu — CHANGES_REQUESTED** [review]

**Review by Kitzunu — APPROVED** [review]

**Comment by Maelthyrr** [comment]
Seems to work fine (tested in Scarlet dungeon and in WSG). No issues find.

<details>
WoWScrnShot_050522_184638
</details>

**Pull request merged**